### PR TITLE
xds: transport level re-connection

### PR DESCRIPTION
**Pull request by Shikugawa**
Description: recreate https://github.com/envoyproxy/envoy/pull/9676
Risk Level: mid
Testing: Unit / Integration
Docs Changes: Required
Release Notes: Required
[Optional Fixes #Issue] #9617 
[Optional Deprecated:] None

### CODE REVIEW

**Comment by lambdai** [comment]
Early comment: have you considered moving updateCluster callback from xDSApiImpl to grpcMux? IMHO LdsApi should react the same regardless it's connection reset or cluster reset

**Comment by mattklein123** [comment]
cc @wgallagher who I would also like to review. Note that @wgallagher is currently trying to merge the delta/SoTW xDS code and IMO we need to let him do that without adding further changes that complicate the process.

**Comment by Shikugawa** [comment]
I will add this functionality after finishing the unification of Delta/SoTW codes.

**Comment by Shikugawa** [comment]
\nostale

**Comment by Shikugawa** [comment]
\nostale

File: include/envoy/config/subscription_factory.h [hunk 1]
```diff
@@ -19,12 +19,13 @@ class SubscriptionFactory {
...
    * @param callbacks the callbacks needed by all Subscription objects, to deliver config updates.
    *                  The callbacks must not result in the deletion of the Subscription object.
+   * @param cluster_index used to specify index of management server.
    * @return SubscriptionPtr subscription object corresponding for config and type_url.
    */
   virtual SubscriptionPtr
   subscriptionFromConfigSource(const envoy::config::core::v3::ConfigSource& config,
                                absl::string_view type_url, Stats::Scope& scope,
-                               SubscriptionCallbacks& callbacks) PURE;
+                               SubscriptionCallbacks& callbacks, int cluster_index) PURE;
```
Comment (htuch) [inline include/envoy/config/subscription_factory.h]: This seems like some kind of layering violation. Shouldn't need to plumb raw integers around abstract interfaces like the subscription factory.

File: include/envoy/config/subscription.h [hunk 2]
```diff
@@ -67,6 +67,11 @@ class SubscriptionCallbacks {
    * a RouteConfiguration, based on the underlying resource type.
    */
   virtual std::string resourceName(const ProtobufWkt::Any& resource) PURE;
+
+  /**
+   * Update cluster of management server.
```
Comment (htuch) [inline include/envoy/config/subscription.h]: What does it means to update the cluster of the management server? Update it to what? There are no params..

File: source/common/config/grpc_mux_impl.cc [hunk 3]
```diff
@@ -222,10 +224,17 @@ void GrpcMuxImpl::onStreamEstablished() {
...
 void GrpcMuxImpl::onEstablishmentFailure() {
-  for (const auto& api_state : api_state_) {
-    for (auto watch : api_state.second.watches_) {
-      watch->callbacks_.onConfigUpdateFailed(
-          Envoy::Config::ConfigUpdateFailureReason::ConnectionFailure, nullptr);
+  // Try to connect all of subscribed clusters which are included on grpc_service. If all of
+  // attempts were failed, It may cause EnvoyException and notify all watches that connection
+  // can't be established.
+  try {
+    callbacks_.updateCluster();
```
Comment (htuch) [inline source/common/config/grpc_mux_impl.cc]: How does this interact with retry timers? Also, onRemoteClose() seems to call into this method, but we might legitimately want to retry on the existing gRPC service.

File: source/common/config/grpc_mux_impl.cc [hunk 4]
```diff
@@ -222,10 +224,17 @@ void GrpcMuxImpl::onStreamEstablished() {
 }
 
 void GrpcMuxImpl::onEstablishmentFailure() {
-  for (const auto& api_state : api_state_) {
-    for (auto watch : api_state.second.watches_) {
-      watch->callbacks_.onConfigUpdateFailed(
-          Envoy::Config::ConfigUpdateFailureReason::ConnectionFailure, nullptr);
+  // Try to connect all of subscribed clusters which are included on grpc_service. If all of
```
Comment (htuch) [inline source/common/config/grpc_mux_impl.cc]: This comment is a bit vague. I think you're trying to suggest that we're actually going to pick the gRPC service in the event of failure and try to connect.

File: source/common/config/grpc_mux_impl.cc [hunk 5]
```diff
@@ -222,10 +224,17 @@ void GrpcMuxImpl::onStreamEstablished() {
...
+  // attempts were failed, It may cause EnvoyException and notify all watches that connection
+  // can't be established.
+  try {
+    callbacks_.updateCluster();
+  } catch (EnvoyException&) {
```
Comment (htuch) [inline source/common/config/grpc_mux_impl.cc]: Why shouldn't we fall back to the 3rd or 4th, etc. gRPC service here?

**Pull request closed without merging**